src/aircloak/JsonApi/JsonApiClient.cs

```diff
@@ -104,137 +97,108 @@ public async Task<DataSourceCollection> GetDataSources()
 
-        /// <summary>
-        /// Polls for a query's results until query resolution is complete, or until a specified timeout.
-        /// </summary>
-        /// <param name="queryId">The query Id obtained via a previous call to the /api/query endpoint.</param>
-        /// <param name="querySpec">An instance of the <see cref="IQuerySpec{TRow}"/> interface.</param>
-        /// <param name="timeout">How long to wait for the query to complete.</param>
-        /// <param name="pollFrequency">Optional. How often to poll the api endpoint. Defaults to
-        /// DefaultPollingFrequencyMillis.</param>
-        /// <typeparam name="TRow">The type to use to deserialise each row returned in the query results.</typeparam>
-        /// <returns>A QueryResult instance. If the query has finished executing, contains the query results, with each
-        /// row seralised to type <c>TRow</c>.</returns>
-        public async Task<QueryResult<TRow>> PollQueryUntilCompleteOrTimeout<TRow>(
-            string queryId,
-            IQuerySpec<TRow> querySpec,
-            TimeSpan timeout,
-            TimeSpan? pollFrequency = null)
-        {
-            using var tokenSource = new CancellationTokenSource();
-            tokenSource.CancelAfter(timeout);
+                    if (queryResult.Query.Completed)
+                    {
+                        switch (queryResult.Query.QueryState)
+                        {
+                            case "completed":
+                                return queryResult;
+                            case "error":
+                                throw new Exception("Aircloak API query error.\n" +
+                                    GetQueryResultDetails(querySpec, queryResult));
+                            case "cancelled":
+                                throw new OperationCanceledException("Aircolak API query canceled.\n" +
+                                    GetQueryResultDetails(querySpec, queryResult));
+                        }
+                    }
 
-            try
-            {
-                return await PollQueryUntilComplete<TRow>(queryId, querySpec, tokenSource.Token, pollFrequency);
+                    cancellationToken.ThrowIfCancellationRequested();
+                    await Task.Delay(pollFrequency, cancellationToken);
+                    cancellationToken.ThrowIfCancellationRequested();
+                }
             }
-            catch (OperationCanceledException e)
+            catch (OperationCanceledException ex)
             {
-                if (e.CancellationToken == tokenSource.Token)
+                if (ex.CancellationToken == cancellationToken)
                 {
-                    throw new TimeoutException($"Timed out while waiting for query results for {queryId}");
+                    await CancelQuery(queryId, cancellationToken);
                 }
-
                 throw;
             }
```

Comment:
I actually modified the code to cancel the Aircloak API query in case of any exception. 
I am not sure the IsCancellationRequested is actually helpful, because the cancellation request can be also detected by the methods called from PollQueryUntilComplete, so we will just get an OperationCancelledException, without the chance of checking the IsCancellationRequested outside of a catch block.
da6c5c0bc330a616c508dbd67cedf68898a1d2ec

Reply:
Ok, that makes sense, I assumed that a canceled Task would simply return normally in a canceled state, didn't realise it threw an exception.